### PR TITLE
Improve compilation times on Windows

### PR DESCRIPTION
**Pull request by vittorioromeo**
Thank you for the very useful library!

Few improvements:
- Better header hygiene
- Isolate `windows.h` to `.cpp` whenever possible
- Use `WIN32_LEAN_AND_MEAN`
- Remove unused headers

Tested on Windows with 
```
cmake .. -DCMAKE_BUILD_TYPE=Debug -GNinja -DCMAKE_EXPORT_COMPILE_COMMANDS=1 
  -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_CXX_FLAGS="-ftime-trace -Wall -Wextra -Wpedantic 
  -Wno-ignored-attributes" -DCMAKE_COLOR_DIAGNOSTICS=1 -DCPPTRACE_BUILD_TESTING=1 
  -DCPPTRACE_BUILD_BENCHMARKING=0
```

There's a lot more that can be improved if you are interested.

### CODE REVIEW

File: src/binary/object.hpp
Comment (jeremy-rifkin) [inline src/binary/object.hpp:12]: Can we `#include <cpptrace/cpptrace.hpp>` here instead. My main concern is this falling out of line with the definition there.

File: src/cpptrace.cpp
Comment (jeremy-rifkin) [inline src/cpptrace.cpp:8]: Placement new is used in this file

File: src/from_current.cpp
Comment (jeremy-rifkin) [inline src/from_current.cpp:24]: Good catch

File: src/platform/dbghelp_syminit_manager.hpp
Comment (jeremy-rifkin) [inline src/platform/dbghelp_syminit_manager.hpp:9]: It'd be helpful to add a comment here indicating it's a `HANDLE`, just to make the intent clear

File: src/unwind/unwind.hpp
Comment (jeremy-rifkin) [inline src/unwind/unwind.hpp:4]:
```suggestion
#include <cpptrace/cpptrace.hpp>
```
(for consistency and since the intent is that it's a library header)

File: src/binary/object.cpp
Comment (jeremy-rifkin) [inline src/binary/object.cpp:19]: Do you think it'd be worthwhile to just do `target_compile_definitions(... PRIVATE WIN32_LEAN_AND_MEAN)`?

File: src/utils/common.hpp
Comment (jeremy-rifkin) [inline src/utils/common.hpp:32]: Good catch

File: src/utils/microfmt.cpp
Comment (jeremy-rifkin) [inline src/utils/microfmt.cpp:9]: Minor formatting thing
```suggestion
    std::ostream& get_cout() {
```

File: src/utils/utils.cpp
Comment (jeremy-rifkin) [inline src/utils/utils.cpp:32]:
```suggestion
    void enable_virtual_terminal_processing_if_needed() noexcept {
```

File: src/utils/utils.hpp
Comment (jeremy-rifkin) [inline src/utils/utils.hpp:13]: Placement new is used in this file too

**Review by jeremy-rifkin — COMMENTED** [review]
Thanks for taking the time to contribute! Header hygiene is always appreciated, as is tidying up includes. I definitely agree there's a lot more that can be improved 😄 I have a small handful of comments, mostly straightforward

**Comment by jeremy-rifkin** [comment]
Looks like some CI is failing, unfortunately one of the hard things about developing this library is how much code can't be tested on a single platform. Happy to help however I can.

File: src/binary/object.hpp
Comment (vittorioromeo) [reply src/binary/object.hpp:12]: I measured this change making a noticeable difference via `ClangBuildAnalyzer`. If there is a mismatch between these forward declarations and `cpptrace.hpp`, compilation will fail with an hard error (i.e. you won't silently introduce any UB).

Given that, would you reconsider?

File: src/cpptrace.cpp
Comment (vittorioromeo) [reply src/cpptrace.cpp:8]: Good catch, will reintroduce.

File: src/platform/dbghelp_syminit_manager.hpp
Comment (vittorioromeo) [reply src/platform/dbghelp_syminit_manager.hpp:9]: Good point, will do!

File: src/unwind/unwind.hpp
Comment (vittorioromeo) [reply src/unwind/unwind.hpp:4]: Will fix

File: src/utils/microfmt.cpp
Comment (vittorioromeo) [reply src/utils/microfmt.cpp:9]: Will fix

File: src/utils/utils.hpp
Comment (vittorioromeo) [reply src/utils/utils.hpp:13]: Will fix

File: src/utils/utils.cpp
Comment (vittorioromeo) [reply src/utils/utils.cpp:32]: Will fix

**Comment by vittorioromeo** [comment]
@jeremy-rifkin: could CI be retriggered'

File: src/binary/object.hpp
Comment (jeremy-rifkin) [reply src/binary/object.hpp:12]: I think that’s reasonable, I appreciate you measuring. I had also been thinking of splitting ip the cpptrace.hpp file so one option in the future might be to have a more minimal header providing the type.

**Comment by vittorioromeo** [comment]
@jeremy-rifkin: and again please :) 

If possible, could you make CI not require approval until this PR is merged? I need to iterate a bit to figure out what includes are missing on various platforms.

**Comment by jeremy-rifkin** [comment]
I’ll look to see if I can set it to not need approval, sorry for the iteration hassle

**Comment by vittorioromeo** [comment]
> I’ll look to see if I can set it to not need approval, sorry for the iteration hassle

Seems like approval is still required :(

**Comment by jeremy-rifkin** [comment]
Sorry for the hassle here, I think the setting is correct now. I can take a stab at getting things correct later too.

**Review by jeremy-rifkin — APPROVED** [review]
Thanks again so much for the contribution, taking the time to do this, and putting up with all the iteration pain. I went ahead and added a couple commits to get remaining stuff working. Everything here LGTM, thanks again!

**Pull request merged**
